ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -372,12 +391,24 @@ class SimpleOperationTracker implements OperationTracker {
         && numActiveReplicasInOriginatingDc >= routerConfig.routerPutSuccessTarget) {
       // There are two conditions to meet in order to use this feature
       // 1. TerminateOnNotFound has to be enabled.
-      // 2. We need enough active replicas in the originating DC
+      // 2. We also need to make sure that there are sufficient number of replicas in leader or stand by state. This is
+      //   because it is possible that local replicas became unavailable due to incorrect batch OS updates, etc. In
+      //   such cases, we shouldn't terminate our search locally and try remote colos.
       // How many active replicas is considered as enough? We have to consider some extreme cases.
       // When the put success target is 2, we at least need 2 replicas in active state so we know the replicas are up to
       // date. If we have only one replica in active state, we might run into a scenario that we are getting NotFound
       // from the other two inactive replicas and we should not ignore the active one.
-      originatingDcNotFoundFailureThreshold = originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1;
+
+      // Count the number of replicas from originating dc that will be used for this operation and set the failure
+      // threshold for NOT_FOUND errors based on it.
+      int originatingDcOperationReplicaCount = 0;
+      for (ReplicaId replicaId : replicaPool) {
+        if (replicaId.getDataNodeId().getDatacenterName().equals(originatingDcName)) {
+          originatingDcOperationReplicaCount++;
+        }
+      }
+      originatingDcNotFoundFailureThreshold =
+          originatingDcOperationReplicaCount - routerConfig.routerPutSuccessTarget + 1;
```

Comment:
I feel we should use "original DC total replica count" instead of "original DC operation replica count".
This threshold means that we have at least 2 Put in the local DC. So if local DC returns 2 NOT_FOUND, we think the blob doesn't exist.
After the change, in the local DC, if one replica is not in the operational state.  I assume the threshold will be 2-2+1 = 1. 
So as long as there is one local DC returned NOT_FOUND, we'll terminate it as NOT_FOUND. But it's possible the "not operational replica" has the Blob, just somehow it's not available at this moment, right? In this case, I think we shouldn't terminate it. We need check with remote replicas, right?